### ISSUE TITLE
Why are we trying to find alpha for which the testing error is maximum? (DET)

### ISSUE THREAD

**Issue by SuvarshaChennareddy** [mlpack/mlpack#3238]
This is a question (potential bug fix) regarding DET's implementation, specifically the part to find an optimal `alpha`.
I'm not sure if I'm missing something here (close this issue if there is indeed nothing wrong), but I can't seem to figure out why we are trying to find `alpha` for which the testing error is **maximum**.  The [code snippet](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/det/dt_utils_impl.hpp#L277) is pasted below:
```cpp
for (size_t i = 0; i < prunedSequence.size() - 1; ++i)
  {
    // We can no longer work in the log-space for this because we have no
    // guarantee the quantity will be positive.
    long double thisError = -std::exp((long double) prunedSequence[i].second) +
        (long double) regularizationConstants[i];

    if (thisError > cvBestError)
    {
      cvBestError = thisError;
      optimalAlpha = prunedSequence[i].first;
    }
  }
```

**Comment by rcurtin** [comment]
Thanks for pointing this out!  Talk about digging in deep :smile: 

I spent some time with the code and realized that actually what is being computed here is the *negative* error.  `prunedSequence[i].second` comes from the function `DTree::LogNegError()`, and I verified that that indeed computes the negative error according to [the paper](https://home.ipipan.waw.pl/j.mielniczuk/Density_estimation_trees.pdf) (see eqn. 6).  However, `thisError` appears to then invert it again.  That line is, basically, `-exp(log(-error) + log(regularization))`, which simplifies to `error * regularization`---but that still should be something we *minimize*, not maximize.  

I went back to @rithram's original code from 12 years ago, and that had a much simpler loop over `prunedSequence` that maximized the negative error.  I kept digging and found that [this change](https://github.com/mlpack/mlpack/commit/fe7d1039a1dd2b893f8f37e524e85ced6ad75b2d) from me in 2012 seems to have inverted the behavior, so that it maximized the error (not the negative error).  The motivation of that change was to make all of the calculations in log-space, and only revert back to 'normal' space at the last possible moment.  However, there is a lot of negation going on, and it looks like I dropped one.

I went through the computation and verified it against the original, and I opened PR #3243, which I think makes the computation correct.

Great find!  Thank you for pointing this out.

**Comment by SuvarshaChennareddy** [comment]
No problem! Yes, the reason I raised this issue is because, while I was reading the code, I couldn't figure out why the negative of the negative error (which is just the error) was being maximized. 
>That line is, basically, -exp(log(-error) + log(regularization)), which simplifies to error * regularization---but that still should be something we minimize, not maximize.

This highlights my thought process.
Anyway, thank you for taking time off to look into this!

**Comment by rcurtin** [comment]
No problem! It was fun to dig into the history here. :) If you find anything more like this, feel free to open more issues!  Bugs like this one are often hard to find because of the difficulty of testing machine learning algorithms: often it is easy to test that the algorithm is not horribly wrong, but subtle issues that reduce accuracy or performance by a small amount can be very hard to test for.

**Comment by SuvarshaChennareddy** [comment]
Sure thing! Just one question: Next time, may I open up a PR instead of opening a new issue if the task is simple enough?

**Comment by SuvarshaChennareddy** [comment]
> That line is, basically, `-exp(log(-error) + log(regularization))`, which simplifies to `error * regularization`---but that still should be something we _minimize_, not maximize.
Also I think its `-exp(log(-error)) + regularization` which simplifies to `error + regularization`. Correct me if I'm wrong.

**Comment by rcurtin** [comment]
Right, exactly, thanks for the correction.  And yeah, feel free to open a PR if you find more things like this, but don't feel obligated to!  Even the bug report itself is very helpful. :+1:

**Comment by SuvarshaChennareddy** [comment]
Thank You!

**Comment by rcurtin** [comment]
Fixed in #3243, thanks again @SuvarshaChennareddy! :+1:
